File: simtemp/user/cli/sysfs_config.py
"""
SysfsConfig Module

Handles sysfs configuration interface for the Challenge 2025 temperature sensor.
Provides methods to read and write sensor configuration via sysfs attributes.
"""

import os
import logging
from pathlib import Path
from typing import Dict, Any, Optional

logger = logging.getLogger(__name__)


class SysfsConfig:
    """Configuration interface for simtemp sensor via sysfs."""
    
    def __init__(self, sysfs_base: str = "/sys/class/misc/simtemp"):
        """
        Initialize sysfs configuration interface.
        
        Args:
            sysfs_base: Base path to sysfs attributes
        """
        self.sysfs_base = Path(sysfs_base)
        self.attributes = {
            'sampling_ms': 'sampling_ms',
            'threshold_mC': 'threshold_mC', 
            'mode': 'mode',
            'stats': 'stats'
        }
        
    def _get_attribute_path(self, attr_name: str) -> Path:
        """Get the full path to a sysfs attribute."""
        if attr_name not in self.attributes:
            raise ValueError(f"Unknown attribute: {attr_name}")
        return self.sysfs_base / self.attributes[attr_name]
    
    def _read_attribute(self, attr_name: str) -> Optional[str]:
        """
        Read a sysfs attribute value.
        
        Args:
            attr_name: Name of the attribute
            
        Returns:
            Attribute value as string or None if error
        """
        try:
            attr_path = self._get_attribute_path(attr_name)
            if not attr_path.exists():
                logger.warning(f"Attribute {attr_name} does not exist at {attr_path}")
                return None
            
            with open(attr_path, 'r') as f:
                value = f.read().strip()
            
            logger.debug(f"Read {attr_name}: {value}")
            return value
            
        except Exception as e:
            logger.error(f"Failed to read attribute {attr_name}: {e}")
            return None
    
    def _write_attribute(self, attr_name: str, value: str) -> bool:
        """
        Write a sysfs attribute value.
        
        Args:
            attr_name: Name of the attribute
            value: Value to write as string
            
        Returns:
            True if successful, False otherwise
        """
        try:
            attr_path = self._get_attribute_path(attr_name)
            if not attr_path.exists():
                logger.error(f"Attribute {attr_name} does not exist at {attr_path}")
                return False
            
            with open(attr_path, 'w') as f:
                f.write(str(value))
            
            logger.debug(f"Wrote {attr_name}: {value}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to write attribute {attr_name}: {e}")
            return False
    
    def get_sampling_period(self) -> Optional[int]:
        """
        Get current sampling period in milliseconds.
        
        Returns:
            Sampling period in ms or None if error
        """
        value = self._read_attribute('sampling_ms')
        if value is not None:
            try:
                return int(value)
            except ValueError:
                logger.error(f"Invalid sampling period value: {value}")
        return None
    
    def set_sampling_period(self, period_ms: int) -> bool:
        """
        Set sampling period in milliseconds.
        
        Args:
            period_ms: Sampling period in milliseconds
            
        Returns:
            True if successful, False otherwise
        """
        if period_ms < 1 or period_ms > 60000:  # 1ms to 60s
            logger.error(f"Invalid sampling period: {period_ms} ms")
            return False
        
        return self._write_attribute('sampling_ms', str(period_ms))
    
    def get_threshold(self) -> Optional[int]:
        """
        Get current threshold in milli-degrees Celsius.
        
        Returns:
            Threshold in milli-degrees Celsius or None if error
        """
        value = self._read_attribute('threshold_mC')
        if value is not None:
            try:
                return int(value)
            except ValueError:
                logger.error(f"Invalid threshold value: {value}")
        return None
    
    def set_threshold(self, threshold_mc: int) -> bool:
        """
        Set threshold in milli-degrees Celsius.
        
        Args:
            threshold_mc: Threshold in milli-degrees Celsius
            
        Returns:
            True if successful, False otherwise
        """
        # Reasonable temperature range: -50°C to 150°C
        if threshold_mc < -50000 or threshold_mc > 150000:
            logger.error(f"Invalid threshold: {threshold_mc} milli-degrees C")
            return False
        
        return self._write_attribute('threshold_mC', str(threshold_mc))
    
    def get_mode(self) -> Optional[str]:
        """
        Get current sensor mode.
        
        Returns:
            Sensor mode ('normal', 'noisy', 'ramp') or None if error
        """
        return self._read_attribute('mode')
    
    def set_mode(self, mode: str) -> bool:
        """
        Set sensor mode.
        
        Args:
            mode: Sensor mode ('normal', 'noisy', 'ramp')
            
        Returns:
            True if successful, False otherwise
        """
        valid_modes = ['normal', 'noisy', 'ramp']
        if mode not in valid_modes:
            logger.error(f"Invalid mode: {mode}. Valid modes: {valid_modes}")
            return False
        
        return self._write_attribute('mode', mode)
    
    def get_stats(self) -> Dict[str, Any]:
        """
        Get sensor statistics.
        
        Returns:
            Dictionary with statistics or empty dict if error
        """
        stats_str = self._read_attribute('stats')
        if stats_str is None:
            return {}
        
        # Parse stats format: "key1:value1,key2:value2,..."
        stats = {}
        try:
            for item in stats_str.split(','):
                if ':' in item:
                    key, value = item.split(':', 1)
                    key = key.strip()
                    value = value.strip()
                    
                    # Try to convert to int, fall back to string
                    try:
                        stats[key] = int(value)
                    except ValueError:
                        stats[key] = value
            
        except Exception as e:
            logger.error(f"Failed to parse stats: {e}")
            return {}
        
        return stats
    
    def get_current_config(self) -> Dict[str, Any]:
        """
        Get all current configuration values.
        
        Returns:
            Dictionary with current configuration
        """
        config = {}
        
        sampling_ms = self.get_sampling_period()
        if sampling_ms is not None:
            config['sampling_ms'] = sampling_ms
        
        threshold_mc = self.get_threshold()
        if threshold_mc is not None:
            config['threshold_mC'] = threshold_mc
        
        mode = self.get_mode()
        if mode is not None:
            config['mode'] = mode
        
        return config
    
    def set_config(self, **kwargs) -> Dict[str, bool]:
        """
        Set multiple configuration values.
        
        Args:
            **kwargs: Configuration parameters (sampling_ms, threshold_mC, mode)
            
        Returns:
            Dictionary with results for each parameter
        """
        results = {}
        
        if 'sampling_ms' in kwargs:
            results['sampling_ms'] = self.set_sampling_period(kwargs['sampling_ms'])
        
        if 'threshold_mC' in kwargs:
            results['threshold_mC'] = self.set_threshold(kwargs['threshold_mC'])
        
        if 'mode' in kwargs:
            results['mode'] = self.set_mode(kwargs['mode'])
        
        return results
    
    def validate_sysfs_access(self) -> Dict[str, bool]:
        """
        Validate access to all sysfs attributes.
        
        Returns:
            Dictionary with access status for each attribute
        """
        access_status = {}
        
        for attr_name in self.attributes:
            attr_path = self._get_attribute_path(attr_name)
            
            exists = attr_path.exists()
            readable = False
            writable = False
            
            if exists:
                readable = os.access(attr_path, os.R_OK)
                # Don't check write access for read-only attributes like stats
                if attr_name != 'stats':
                    writable = os.access(attr_path, os.W_OK)
                else:
                    writable = True  # Stats is read-only, so mark as "writable" (accessible)
            
            access_status[attr_name] = {
                'exists': exists,
                'readable': readable,
                'writable': writable,
                'path': str(attr_path)
            }
        
        return access_status
    
    def reset_to_defaults(self) -> bool:
        """
        Reset sensor configuration to default values.
        
        Returns:
            True if all defaults were set successfully
        """
        defaults = {
            'sampling_ms': 100,      # 100ms default sampling
            'threshold_mC': 45000,   # 45°C default threshold
            'mode': 'normal'         # Normal mode
        }
        
        results = self.set_config(**defaults)
        success = all(results.values())
        
        if success:
            logger.info("Reset sensor configuration to defaults")
        else:
            logger.warning(f"Failed to reset some configuration values: {results}")
        
        return success
    
    def __str__(self) -> str:
        """String representation of the configuration."""
        config = self.get_current_config()
        stats = self.get_stats()
        
        result = f"SysfsConfig({self.sysfs_base})\n"
        result += "Configuration:\n"
        for key, value in config.items():
            result += f"  {key}: {value}\n"
        
        if stats:
            result += "Statistics:\n"
            for key, value in stats.items():
                result += f"  {key}: {value}\n"
        
        return result


# Convenience functions
def get_sensor_config(sysfs_base: str = "/sys/class/misc/simtemp") -> Dict[str, Any]:
    """
    Get current sensor configuration.
    
    Args:
        sysfs_base: Base path to sysfs attributes
        
    Returns:
        Current configuration dictionary
    """
    config = SysfsConfig(sysfs_base)
    return config.get_current_config()


def set_sensor_config(sysfs_base: str = "/sys/class/misc/simtemp", **kwargs) -> bool:
    """
    Set sensor configuration parameters.
    
    Args:
        sysfs_base: Base path to sysfs attributes
        **kwargs: Configuration parameters
        
    Returns:
        True if all parameters were set successfully
    """
    config = SysfsConfig(sysfs_base)
    results = config.set_config(**kwargs)
    return all(results.values())


if __name__ == "__main__":
    # Simple test when run directly
    logging.basicConfig(level=logging.INFO)
    
    print("Testing SysfsConfig...")
    config = SysfsConfig()
    
    print("Validating sysfs access...")
    access = config.validate_sysfs_access()
    for attr, status in access.items():
        print(f"  {attr}: exists={status['exists']}, readable={status['readable']}, writable={status['writable']}")
    
    print("\nCurrent configuration:")
    current = config.get_current_config()
    for key, value in current.items():
        print(f"  {key}: {value}")
    
    print("\nCurrent statistics:")
    stats = config.get_stats()
    for key, value in stats.items():
        print(f"  {key}: {value}")

File: simtemp/user/cli/sysfs_config.c
/* SPDX-License-Identifier: GPL-2.0+ */
/*
 * SimTemp CLI - Sysfs Configuration
 *
 * Copyright (c) 2025 Jorge Rodriguez Moreno
 *
 * Sysfs attribute access functions for configuring the SimTemp
 * temperature sensor driver parameters.
 */

#include "simtemp_cli.h"

int read_sysfs_attribute(const char *base_path, const char *attr_name, 
                        char *value, size_t value_size)
{
    char path[512];
    FILE *file;
    char *newline;
    
    /* Build full attribute path */
    snprintf(path, sizeof(path), "%s/%s", base_path, attr_name);
    
    /* Open attribute file */
    file = fopen(path, "r");
    if (!file) {
        print_error("Failed to open sysfs attribute %s: %s", path, strerror(errno));
        return -1;
    }
    
    /* Read value */
    if (!fgets(value, value_size, file)) {
        print_error("Failed to read sysfs attribute %s: %s", path, strerror(errno));
        fclose(file);
        return -1;
    }
    
    fclose(file);
    
    /* Remove trailing newline */
    newline = strchr(value, '\n');
    if (newline) {
        *newline = '\0';
    }
    
    return 0;
}

int write_sysfs_attribute(const char *base_path, const char *attr_name, 
                         const char *value)
{
    char path[512];
    FILE *file;
    
    /* Build full attribute path */
    snprintf(path, sizeof(path), "%s/%s", base_path, attr_name);
    
    /* Open attribute file for writing */
    file = fopen(path, "w");
    if (!file) {
        print_error("Failed to open sysfs attribute %s for writing: %s", 
                   path, strerror(errno));
        return -1;
    }
    
    /* Write value */
    if (fprintf(file, "%s", value) < 0) {
        print_error("Failed to write sysfs attribute %s: %s", path, strerror(errno));
        fclose(file);
        return -1;
    }
    
    fclose(file);
    
    return 0;
}

int get_sampling_period(const char *sysfs_base)
{
    char value[64];
    
    if (read_sysfs_attribute(sysfs_base, "sampling_ms", value, sizeof(value)) < 0) {
        return -1;
    }
    
    return atoi(value);
}

int set_sampling_period(const char *sysfs_base, int period_ms)
{
    char value[64];
    
    if (!validate_sampling_period(period_ms)) {
        print_error("Invalid sampling period: %d ms (valid range: 1-60000)", period_ms);
        return -1;
    }
    
    snprintf(value, sizeof(value), "%d", period_ms);
    
    return write_sysfs_attribute(sysfs_base, "sampling_ms", value);
}

int get_threshold(const char *sysfs_base)
{
    char value[64];
    
    if (read_sysfs_attribute(sysfs_base, "threshold_mC", value, sizeof(value)) < 0) {
        return -1;
    }
    
    return atoi(value);
}

int set_threshold(const char *sysfs_base, int threshold_mc)
{
    char value[64];
    
    if (!validate_threshold(threshold_mc)) {
        print_error("Invalid threshold: %d mC (valid range: -50000-150000)", threshold_mc);
        return -1;
    }
    
    snprintf(value, sizeof(value), "%d", threshold_mc);
    
    return write_sysfs_attribute(sysfs_base, "threshold_mC", value);
}

int get_mode(const char *sysfs_base, char *mode, size_t mode_size)
{
    return read_sysfs_attribute(sysfs_base, "mode", mode, mode_size);
}

int set_mode(const char *sysfs_base, const char *mode)
{
    if (!validate_mode(mode)) {
        print_error("Invalid mode: %s (valid modes: normal, noisy, ramp, static, linear)", mode);
        return -1;
    }
    
    return write_sysfs_attribute(sysfs_base, "mode", mode);
}

int configure_sensor(const struct cli_config *config, struct remote_context *ctx)
{
    int result = 0;
    bool any_config = false;
    
    if (is_remote_connection(config)) {
        /* Remote configuration */
        print_verbose(config, "Configuring remote sensor via %s", 
                     get_connection_type_string(config->conn_type));
        
        if (config->sampling_ms > 0) {
            if (remote_set_sampling_period(ctx, config->sysfs_base, config->sampling_ms) < 0) {
                print_error("Failed to set remote sampling period");
                result = -1;
            } else {
                printf("Set remote sampling period to %d ms\n", config->sampling_ms);
                any_config = true;
            }
        }
        
        if (config->threshold_mc > 0) {
            if (remote_set_threshold(ctx, config->sysfs_base, config->threshold_mc) < 0) {
                print_error("Failed to set remote threshold");
                result = -1;
            } else {
                printf("Set remote threshold to %.1f°C\n", 
                       millicelsius_to_celsius(config->threshold_mc));
                any_config = true;
            }
        }
        
        if (strlen(config->mode) > 0 && strcmp(config->mode, "normal") != 0) {
            if (remote_write_sysfs_attribute(ctx, config->sysfs_base, "mode", config->mode) < 0) {
                print_error("Failed to set remote mode");
                result = -1;
            } else {
                printf("Set remote mode to %s\n", config->mode);
                any_config = true;
            }
        }
    } else {
        /* Local configuration */
        print_verbose(config, "Configuring local sensor");
        
        if (config->sampling_ms > 0) {
            if (set_sampling_period(config->sysfs_base, config->sampling_ms) < 0) {
                print_error("Failed to set sampling period");
                result = -1;
            } else {
                printf("Set sampling period to %d ms\n", config->sampling_ms);
                any_config = true;
            }
        }
        
        if (config->threshold_mc > 0) {
            if (set_threshold(config->sysfs_base, config->threshold_mc) < 0) {
                print_error("Failed to set threshold");
                result = -1;
            } else {
                printf("Set threshold to %.1f°C\n", 
                       millicelsius_to_celsius(config->threshold_mc));
                any_config = true;
            }
        }
        
        if (strlen(config->mode) > 0 && strcmp(config->mode, "normal") != 0) {
            if (set_mode(config->sysfs_base, config->mode) < 0) {
                print_error("Failed to set mode");
                result = -1;
            } else {
                printf("Set mode to %s\n", config->mode);
                any_config = true;
            }
        }
    }
    
    if (!any_config) {
        printf("No configuration parameters specified. Use --help for options.\n");
        return show_sensor_status(config, ctx);
    }
    
    return result;
}

int show_sensor_status(const struct cli_config *config, struct remote_context *ctx)
{
    int sampling_ms, threshold_mc;
    char mode[64];
    char stats[512];
    
    printf("\n=== SimTemp Sensor Status ===\n");
    printf("Connection: %s\n", get_connection_type_string(config->conn_type));
    
    if (is_remote_connection(config)) {
        printf("Remote host: %s:%d\n", config->remote_host, config->remote_port);
        printf("Username: %s\n", config->username);
    }
    
    printf("Device path: %s\n", config->device_path);
    printf("Sysfs path: %s\n", config->sysfs_base);
    
    /* Read current configuration */
    if (is_remote_connection(config)) {
        /* Remote status */
        sampling_ms = remote_get_sampling_period(ctx, config->sysfs_base);
        threshold_mc = remote_get_threshold(ctx, config->sysfs_base);
        
        if (remote_read_sysfs_attribute(ctx, config->sysfs_base, "mode", mode, sizeof(mode)) < 0) {
            strcpy(mode, "unknown");
        }
        
        if (remote_read_sysfs_attribute(ctx, config->sysfs_base, "stats", stats, sizeof(stats)) < 0) {
            strcpy(stats, "unavailable");
        }
    } else {
        /* Local status */
        sampling_ms = get_sampling_period(config->sysfs_base);
        threshold_mc = get_threshold(config->sysfs_base);
        
        if (get_mode(config->sysfs_base, mode, sizeof(mode)) < 0) {
            strcpy(mode, "unknown");
        }
        
        if (read_sysfs_attribute(config->sysfs_base, "stats", stats, sizeof(stats)) < 0) {
            strcpy(stats, "unavailable");
        }
    }
    
    printf("\n=== Configuration ===\n");
    if (sampling_ms >= 0) {
        printf("Sampling period: %d ms\n", sampling_ms);
    } else {
        printf("Sampling period: unavailable\n");
    }
    
    if (threshold_mc >= -200000) {  /* Check for reasonable value */
        printf("Threshold: %.1f°C (%d mC)\n", 
               millicelsius_to_celsius(threshold_mc), threshold_mc);
    } else {
        printf("Threshold: unavailable\n");
    }
    
    printf("Mode: %s\n", mode);
    
    printf("\n=== Statistics ===\n");
    if (strcmp(stats, "unavailable") != 0) {
        /* Parse and display stats */
        char *token = strtok(stats, ",");
        while (token) {
            char *colon = strchr(token, ':');
            if (colon) {
                *colon = '\0';
                printf("%s: %s\n", token, colon + 1);
            }
            token = strtok(NULL, ",");
        }
    } else {
        printf("Statistics unavailable\n");
    }
    
    printf("\n");
    
    return 0;
}